src/mlpack/core/util/param.hpp

```diff
@@ -15,6 +15,24 @@
 #ifndef MLPACK_CORE_UTIL_PARAM_HPP
 #define MLPACK_CORE_UTIL_PARAM_HPP
 
+#include <string>
+
+// PROGRAM_NAME is used for mlpackMain modification (mlpackMain##PROGRAM_NAME)
+// PROGRAM_NAME_SUBSTUTIDE is used for passing unique program name to TestOption
+#ifdef PROGRAM_NAME
+#define PROGRAM_NAME_SUBSTITUDE PROGRAM_NAME
+#else
+#define PROGRAM_NAME
+#define PROGRAM_NAME_SUBSTITUDE programNameSubstitude
+static const std::string programNameSubstitude = "programNameSubstitude";
+#endif
+// The MAIN macro is mlpackMain##PROGRAM_NAME. The goal is to have different
+// names of main procedure for main tests when all method's  main functions
+// are linked into one executable (mlpack_test)
+#define TOKENPASTE1(x, y) x ## y
+#define TOKENPASTE2(x, y) TOKENPASTE1(x, y)
+#define MAIN TOKENPASTE2(mlpackMain, PROGRAM_NAME)
```

Comment:
If we can declare `mlpackMain()` as `inline` I think we can simplify this greatly... let me know what you think if we adapted this to the following (or something like it---I haven't tested what is below):

```
// PROGRAM_NAME might be defined for some binding types and may need to be passed
// to the individual parameters.  If it isn't defined, we'll pass a dummy empty string to the
// parameter constructor.
#ifndef PROGRAM_NAME
#define PROGRAM_NAME ""
#endif
```